File: backend/app/schemas/cycle_count.py
"""Explicit public count workspace contracts; timestamps are served in UTC."""

from datetime import date, datetime
from typing import List, Optional

from app.schemas.base import UTCModel


class CycleCountSummary(UTCModel):
    id: int
    count_number: str
    status: str
    scheduled_date: date
    started_at: Optional[datetime]
    completed_at: Optional[datetime]
    warehouse: Optional[str]
    location_code: Optional[str]
    part_id: Optional[int]
    assigned_to: Optional[int]
    assigned_to_name: Optional[str]
    total_items: int
    items_counted: int
    items_adjusted: int
    total_variance_value: float
    notes: Optional[str]


class CycleCountLine(UTCModel):
    id: int
    inventory_item_id: int
    part_id: Optional[int]
    part_number: str
    part_name: str
    unit_of_measure: str
    location: Optional[str]
    lot_number: Optional[str]
    serial_number: Optional[str]
    system_quantity: float
    current_quantity: Optional[float]
    counted_quantity: Optional[float]
    variance: Optional[float]
    variance_value: Optional[float]
    posting_delta: float
    stock_changed: bool
    is_counted: bool
    requires_recount: bool
    counted_at: Optional[datetime]
    notes: Optional[str]


class CycleCountDetail(CycleCountSummary):
    items: List[CycleCountLine]


class CycleCountReview(CycleCountDetail):
    review_token: str


class CycleCountWorkspace(UTCModel):
    items: List[CycleCountSummary]
    total: int
    has_more: bool


class CycleCounter(UTCModel):
    id: int
    name: str

File: frontend/src/types/cycleCount.ts
export interface CycleCountSummary {
  id: number;
  count_number: string;
  status: 'scheduled' | 'in_progress' | 'completed' | 'cancelled';
  scheduled_date: string;
  started_at: string | null;
  completed_at: string | null;
  warehouse: string | null;
  location_code: string | null;
  part_id: number | null;
  assigned_to: number | null;
  assigned_to_name: string | null;
  total_items: number;
  items_counted: number;
  items_adjusted: number;
  total_variance_value: number;
  notes: string | null;
}

export interface CycleCountLine {
  id: number;
  inventory_item_id: number;
  part_id: number | null;
  part_number: string;
  part_name: string;
  unit_of_measure: string;
  location: string | null;
  lot_number: string | null;
  serial_number: string | null;
  system_quantity: number;
  current_quantity: number | null;
  counted_quantity: number | null;
  variance: number | null;
  variance_value: number | null;
  posting_delta: number;
  stock_changed: boolean;
  is_counted: boolean;
  requires_recount: boolean;
  counted_at: string | null;
  notes: string | null;
}

export interface CycleCountDetail extends CycleCountSummary {
  items: CycleCountLine[];
}
export interface CycleCountReview extends CycleCountDetail {
  review_token: string;
}
export interface CycleCountCreate {
  scheduled_date: string;
  warehouse?: string;
  location_code?: string;
  part_id?: number;
  assigned_to?: number;
  notes?: string;
}
